Example of Fluent Bit with EKS, MSK, and Opensearch

priyavartk opened the pull request:
*Issue #, if available (include [keywords](https://help.github.com/articles/closing-issues-using-keywords/) to close issue as applicable, e.g. "fixes <##>"):*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
add eks-msk-opensearch example

Commit:
update Readme.md

Commit:
update Readme.md

Commit:
update Readme.md

Commit:
Add Ref arch Diagram

Commit:
Update Readme

Commit:
Update Readme

Commit:
Remove Zone-id variable, Connector is not able to resovle private hosted zones

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Delete README.md

priyavartk committed:
Rename 0-provider.tf to provider.tf

priyavartk committed:
Rename 1-locals.tf to locals.tf

priyavartk committed:
Rename 3-variables.tf to variables.tf

priyavartk committed:
Rename 2-data.tf to data.tf

priyavartk committed:
Rename 4-vpc.tf to vpc.tf

priyavartk committed:
Rename 5-eks.tf to eks.tf

priyavartk committed:
Rename 6-irsa.tf to irsa.tf

priyavartk committed:
Rename 8-sa.tf to sa.tf

priyavartk committed:
Rename 9-namespaces.tf to namespaces.tf

priyavartk committed:
Rename 7-lbc.tf to lbc.tf

priyavartk committed:
Rename 14-output.tf to output.tf

priyavartk committed:
Rename 10-msk.tf to msk.tf

priyavartk committed:
Rename 12-es.tf to es.tf

priyavartk committed:
Rename 13-fluentbit.tf to fluentbit.tf

priyavartk committed:
Rename 11-iam.tf to iam.tf

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Delete Ref-Architecture.png

shahim committed:
Add logging for multi-tenant applications on EKS

shahim committed:
update readme

shahim committed:
Fix opensearch password validation

shahim committed:
Fix use kubelet issue

priyavartk committed:
Merge branch 'mainline' into mainline

priyavartk committed:
Merge pull request #1 from shahim/mainline

Centralized logging for multi-tenant applications on Amazon EKS

Commit:
update fluent-bit config and remove lb controller code

Commit:
remove bucket name

shahim committed:
fix fluent-bit config

priyavartk committed:
Update msk.tf

priyavartk committed:
Update msk.tf

priyavartk committed:
Update variables.tf

priyavartk committed:
Update README.md

priyavartk committed:
Update variables.tf

priyavartk committed:
Update msk.tf

priyavartk committed:
Merge branch 'mainline' into mainline

priyavartk committed:
Merge pull request #2 from shahim/mainline

fix fluent-bit config

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update msk.tf

PettitWesley commented:
Original PR with review history: https://github.com/aws-samples/amazon-eks-fluent-logging-examples/pull/6

PettitWesley commented on examples/eks-msk-opensearch/terraform/iam_policies/logging.json:
```diff
@@ -0,0 +1,12 @@
+        {
+            "Effect": "Allow",
+            "Action": [
+                "kafka:*"
```
is it possible to scope down this policy? Is this used by FLB to put logs?

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,134 @@
+            Match               application.*
+            Regex               $kubernetes['namespace_name'] ${namespaces}
+
+        [FILTER]
+            Name                lua
+            Match               application.*
+            call                set_topic
+            code                function set_topic(tag, timestamp, record) if record["kubernetes"]["namespace_name"] ~= nil then record["namespace"] = "logs_" .. record["kubernetes"]["namespace_name"] end return 2, timestamp, record end
+
```
So lua filters and filters in general do add overhead... I am not sure we truly need this lua filter. 

I understand that the kafka topic_key config requires that the topic key be top level, not nested: https://docs.fluentbit.io/manual/pipeline/outputs/kafka/

However, you can use filter nest to change the namespace_name key to be top level: https://docs.fluentbit.io/manual/pipeline/filters/nest

So the only thing that's special that the lua filter is doing for you is to add the `logs_` prefix. Do we really need that?

If you think we do, I think it'd be nice to still note the nest filter alternate option in the readme somewhere.

priyavartk replied:
The idea of lua script is not fillter on kubernetes namespaces and set topic name, Originally I am of the view that same kafka cluster may be used other purpose  also.

PettitWesley replied:
I think your response does not answer my point here which is that the lua filter is simply adding the `log_` prefix- why is that very important? You could achieve the same isolation on namespace goal by other means as noted above. Could we give both options and let the user decide if the `logs_` prefix is really important?

PettitWesley replied:
Please consider this.

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,134 @@
+            Tag                 application.*
+            Exclude_Path        /var/log/containers/cloudwatch-agent*, /var/log/containers/fluent-bit*, /var/log/containers/aws-node*, /var/log/containers/kube-proxy*
+            Path                /var/log/containers/*.log
+            Docker_Mode         On
+            Docker_Mode_Flush   5
+            Docker_Mode_Parser  container_firstline
```
remove all three of these and change them too:

```
multiline.parser  docker, cri
```

These are old config options. You also then don't need the container_firstline parser in the parsers file. 

https://docs.fluentbit.io/manual/pipeline/inputs/tail#multiline-and-containers-v1.8

priyavartk replied:
Updated

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update logging.json

priyavartk committed:
Update iam.tf

priyavartk committed:
Update logging.json

PettitWesley commented on examples/eks-msk-opensearch/README.md:
```diff
@@ -0,0 +1,162 @@
+
+In this example, we assumed that you use Kubernetes Namespaces to divide cluster resources between multiple users or tenants as a foundation for multi-tenancy. Furthermore, during log ingestion and storage, we used Kafka topics and OpenSearch indices to isolate logs generated by each tenant. This approach enables you to control who can access logs in any stage.
+
+ In the following diagram Product, Order, and Payment are example of tenants which isolated by Namespaces in Kubernetes, Topics in Kafka, and Indices in OpenSearch.
```
nit: space at beginning of paragraph

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,132 @@
+  # We additionally expose the HTTP service so that Terraform can ping it.
+  service: |
+      [SERVICE]
+        Flush                     30
```
Sorry i didn't notice this before. Why is the Flush interval so high? This will not be ideal for many high throughput envs. I'd like to keep the `Flush` as default. 

Also, you need to add `Grace 30` here since the default k8s SIGTERM to SIGKILL timeout is 30.

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,132 @@
+            Tag                 application.*
+            Exclude_Path        /var/log/containers/cloudwatch-agent*, /var/log/containers/fluent-bit*, /var/log/containers/aws-node*, /var/log/containers/kube-proxy*
+            Path                /var/log/containers/*.log
+            multiline.parser    docker, cri
+            Parser              docker
```
with the `multiline.parser    docker, cri` there is no longer any need to also set `Parser              docker`

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,132 @@
+            code                function set_topic(tag, timestamp, record) if record["kubernetes"]["namespace_name"] ~= nil then record["namespace"] = "logs_" .. record["kubernetes"]["namespace_name"] end return 2, timestamp, record end
+
+    parsers.conf: |
+        [PARSER]
+            Name                docker
+            Format              json
+            Time_Key            time
+            Time_Format         %Y-%m-%dT%H:%M:%S.%LZ
+        [PARSER]
+            Name                syslog
+            Format              regex
+            Regex               ^(?<time>[^ ]* {1,2}[^ ]* [^ ]*) (?<host>[^ ]*) (?<ident>[a-zA-Z0-9_\/\.\-]*)(?:\[(?<pid>[0-9]+)\])?(?:[^\:]*\:)? *(?<message>.*)$
+            Time_Key            time
+            Time_Format         %b %d %H:%M:%S
+        [PARSER]
+            Name                container_firstline
+            Format              regex
+            Regex               (?<log>(?<="log":")\S(?!\.).*?)(?<!\\)".*(?<stream>(?<="stream":").*?)".*(?<time>\d{4}-\d{1,2}-\d{1,2}T\d{2}:\d{2}:\d{2}\.\w*).*(?=})
+            Time_Key            time
+            Time_Format         %Y-%m-%dT%H:%M:%S.%LZ
+        [PARSER]
+            Name                cwagent_firstline
```
Once you address my comment above on the `Parser`, I think this entire file is no longer used?

PettitWesley reviewed (changes requested):
If you can address these quickly, we can get this merged

priyavartk committed:
Update fluent-bit.yaml

priyavartk committed:
Update README.md

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,100 @@
+        Match               application.*
+        Brokers             ${kafka_brokers}
+        topics              logs_kubernetes
+        topic_key           ${namespaces}
```
I think this isn't ideal anymore. 

Let's still set the topic as namespace from the logs. 

The topic_key has to be top level IIRC. K8s filter will add it nested. I think you need to use nest filter to move it to top level: https://docs.fluentbit.io/manual/pipeline/filters/nest

I'm also in favor of adding your lua script in the readme as an example of how to fully customize the topic.  We don't need to throw away that work, someone will find it useful I bet.

PettitWesley replied:
k8s metadata will look like:

```
kubernetes: {
    annotations: {
        "kubernetes.io/psp": "eks.privileged"
    },
    container_hash: "<some hash>",
    container_name: "myapp",
    docker_id: "<some id>",
    host: "ip-10-1-128-166.us-east-2.compute.internal",
    labels: {
        app: "myapp",
        "pod-template-hash": "<some hash>"
    },
    namespace_name: "my-namespace",
    pod_id: "198f7dd2-2270-11ea-be47-0a5d932f5920",
    pod_name: "myapp-5468c5d4d7-n2swr"
}
```

So we can lift everything under that key to top level:

```
[FILTER]
    Name nest
    Match *
    Operation lift
    Nested_under kubernetes
    Add_prefix k8s_
```

The Add_prefix is may be not needed, it adds `k8s_` to every key under kubernetes. You could also make this `logs_` if you want...

And then you set:

```
topic_key k8s_namespace_name
```

priyavartk replied:
have checked in with your suggestion  and dropped lua script as a note at the end of ReadMe

priyavartk committed:
Update fluent-bit.yaml

Add Nest filter instead of lua script lift namespaces up under the same key and add_prefix to set custom topic_key with logs_ prefix

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

priyavartk committed:
Update README.md

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,106 @@
+            Match               application.* 
+            Operation           lift 
+            Nested_under        kubernetes
+            Add_prefix          logs
```
I think you should make it `logs_` otherwise you will get keys like `logsnamespace_name`

PettitWesley replied:
I think you should make it `logs_` otherwise you will get keys like `logsnamespace_name`

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,106 @@
+        Match               application.*
+        Brokers             ${kafka_brokers}
+        topics              kubernetes
+        topic_key           logs_namespaces_name
```
As noted above, your prefix needs to be changed to `logs_`

priyavartk committed:
Update fluent-bit.yaml

Fixed prefix
type in topic name , it will be namespace_name not namespaces_name.
Tested

PettitWesley commented on examples/eks-msk-opensearch/terraform/templates/fluent-bit.yaml:
```diff
@@ -0,0 +1,106 @@
+        [FILTER]
+            Name                grep
+            Match               application.*
+            Regex               $kubernetes['namespace_name'] ${namespaces}
```
Sorry I think I didn;t look closely enough at this before. How is this namespaces variable set? Is it set by the namespaces.tf? What does its value look like? Have you tested this? The grep filter expects a regular expression...

priyavartk replied:
yes. tested with multiple namespaces.

variable.tf has a variable called namespaces "variable "namespaces" {"

this contains the value of all namespaces which are sent for logging

thats why I added 2 more namespaces to variable.tf

so the logging namespaces is coming from variables.tf
[
and then fluentbit.tf creates namespaces regex

here namespaces = join("|", [for namespace in var.namespaces : namespace.name if namespace.enable_logs_to_es])

priyavartk replied:
I thought I replied on testing in my previous resolution comments. 
It works.

priyavartk replied:
namespaces.tf is contaiers the name of k8s namespaces to which a user want to enable logging. the grp filter helps to only filter namepsaces which have been set to "true"

PettitWesley reviewed (changes requested):
@priyavartk I need an answer on this bit- have you tested it?

priyavartk commented:
yes. tested with multiple namespaces. 

variable.tf has a variable called namespaces "variable "namespaces" {"

this contains the value of all namespaces which are sent for logging

thats why I added 2 more namespaces to variable.tf

so the logging namespaces is coming from variables.tf

and then fluentbit.tf creates namespaces regex

here namespaces           = join("|", [for namespace in var.namespaces : namespace.name if namespace.enable_logs_to_es])

PettitWesley commented on examples/eks-msk-opensearch/terraform/fluentbit.tf:
```diff
@@ -0,0 +1,48 @@
+      service_account_name = kubernetes_service_account.logging.metadata.0.name,
+      kafka_brokers        = join(",", formatlist("%s", split(",", module.kafka.bootstrap_brokers))),
+
+      namespaces           = join("|", [for namespace in var.namespaces : namespace.name if namespace.enable_logs_to_es])
```
ah I see...
